### PR TITLE
Basic prototype

### PR DESCRIPTION
**Pull request by richfitz**
This PR implements the basic api, we can add more later. It looks large but most lines are json schemas and tests. The implementation is 90 lines apparently

* three endpoints: `GET /`, `POST /validate` and `POST /compile`
* the output of `POST /compile` will definitely need tweaking, most of the work for that is in odin though
* build simple docker image, test that we can bring it up per usual
* json logging from porcelain

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,6 +15,90 @@ To install `odin.api`:
...
+  }
+}
+```
+
+Check a model with `POST /validate`
+
+```
+$ curl -s -H 'Content-Type: application/json' \
+   --data '{"model": "deriv(x) <- a\ninitial(x) <- 1\na <- user(1)"}' \
+   http://localhost:8001/validate | jq
```
Comment (EmmaLRussell) [inline README.md:46]: Might be good to include an example of response when validation fails. Still returns a 200, but schema is different?

File: README.md [hunk 2]
```diff
@@ -15,6 +15,90 @@ To install `odin.api`:
...
+          "default": 1,
+          "min": null,
+          "max": null,
+          "is_integer": false,
+          "rank": 0
+        }
+      ],
+      "messages": []
+    },
+    "model": "...large blob of js removed..."
```
Comment (EmmaLRussell) [inline README.md:119]: Would it be useful to include a bit more of a clue about what the js is going to look like? Is it a function?

**Review by EmmaLRussell — COMMENTED** [review]
Looks good to me, as far as I can tell!

Disappointed it's not called wodinr but this is probably more sensible! :)

File: R/api.R [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+##' @porcelain POST /validate => json(validate_response)
+##'   body data :: json(validate_request)
+model_validate <- function(data) {
+  data <- jsonlite::fromJSON(data, simplifyDataFrame = FALSE)
+  odin_js_validate(data$model)
+}
+
+
+##' @porcelain POST /compile => json(compile_response)
```
Comment (hillalex) [inline R/api.R:39]: hadn't seen porcelain code generating comment attributes before - cool :)

File: README.md [hunk 4]
```diff
@@ -15,6 +15,90 @@ To install `odin.api`:
 remotes::install_github("mrc-ide/odin.api", upgrade = FALSE)
 ```
 
+## Usage
+
+```
+docker pull mrcide/odin.api
+docker run -d --rm -p 8001:8001 mrcide/odin.api
```
Comment (hillalex) [inline README.md]: Might be nice to indicate that you can pull individual branches with `docker pull mrcide/odin.api:branch-name` as well, which is what I did to test this branch.

**Review by hillalex — APPROVED** [review]
Works for me :)

File: README.md (hunk 1 above)
Comment (richfitz) [reply README.md:46]: I'll add one sure. The return code is something we can change if you'd prefer but it feels like it should be 200 (the data sent was fine but model is bad). There's an example in the tests, but will wire one in here (probably should make these automatically generated too)

File: README.md (hunk 2 above)
Comment (richfitz) [reply README.md:119]: This is something that 100% will need changing. The output at present looks like this: https://github.com/mrc-ide/odin.js#use-from-webpages but minified by default. I have no idea what a convenient form would be, and presume that will depend a lot on your frameworks etc.

File: R/api.R (hunk 3 above)
Comment (richfitz) [reply R/api.R:39]: We only just merged this! Pretty happy with how it has worked out here, though when things go wrong the error messages do need improving

**Review by EmmaLRussell — APPROVED** [review]

**Pull request merged**